app/database/build_store.py

```diff
@@ -0,0 +1,520 @@
+                self._session.add(failed_artifact_directory)
+
+        # FailedSubjobAtomPairs
+        if build._build_artifact is not None:
+            for subjob_id, atom_id in build._build_artifact.get_failed_subjob_and_atom_ids():
+                failed_subjob_and_atom_ids = FailedSubjobAtomPairsSchema(
+                    build_id=build_id,
+                    subjob_id=subjob_id,
+                    atom_id=atom_id
+                )
+                self._session.add(failed_subjob_and_atom_ids)
+
+        # BuildRequest
+        build_params = build._build_request._build_parameters
+        build_request = BuildRequestSchema(
+            build_id=build_id,
+            build_parameters=json.dumps(build._build_request.build_parameters())
+        )
+        self._session.add(build_request)
+
+        # BuildFsm
+        fsm_timestamps = {state.lower(): timestamp for state, timestamp in build._state_machine.transition_timestamps.items()}
+        build_fsm = BuildFsmSchema(
+            build_id=build_id,
+            state=build._status(),
+            queued=fsm_timestamps['queued'],
+            finished=fsm_timestamps['finished'],
+            prepared=fsm_timestamps['prepared'],
+            preparing=fsm_timestamps['preparing'],
+            error=fsm_timestamps['error'],
+            canceled=fsm_timestamps['canceled'],
+            building=fsm_timestamps['building']
+        )
+        self._session.add(build_fsm)
+
+        # Subjobs
+        subjobs = build._all_subjobs_by_id
+        for subjob_id in subjobs:
+            subjob = build._all_subjobs_by_id[subjob_id]
+            subjob_schema = SubjobsSchema(
+                subjob_id=subjob_id,
+                build_id=build_id,
+                completed=subjob.completed
+            )
+            self._session.add(subjob_schema)
+
+            # Atoms
+            for atom in subjob._atoms:
+                atom_schema = AtomsSchema(
+                    atom_id=atom.id,
+                    build_id=build_id,
+                    subjob_id=subjob_id,
+                    command_string=atom.command_string,
+                    expected_time=atom.expected_time,
+                    actual_time=atom.actual_time,
+                    exit_code=atom.exit_code,
+                    state=atom.state,
+                )
+                self._session.add(atom_schema)
+
+        # Save changes
+        self._session.commit()
+        return build_id
+
+    def _update_build(self, build: Build) -> int:
+        """
+        Serialize a Build object and update all of the parts to the database.
+        NOTE: These changes are not committed here. If you want these changes to persist,
+              make sure to commit the session afterwards.
+              We do selectively call commit a few times here but only after we delete rows.
+        :param build_id: The build_id of the build to update in the database.
+        """
+        build_id = build.build_id()
+        (
```

Comment:
Not sure what we'd prefer to do with something like this formatting wise. Alternatively the method that returns this tuple could return a dict instead or something